### PR TITLE
fix isValid issue with validationFields and consistency with isValid

### PR DESCRIPTION
**Pull request by bluebill1049**
This PR is about: 

- `isValid` will return false consistently through different validation `mode`
- fix a bug which not includes `validationFields` for `isValid` form state

### CODE REVIEW

File: src/useForm.ts [hunk 1]
```diff
@@ -547,7 +547,8 @@ export default function useForm<FormValues extends FieldValues = FieldValues>({
     }
 
     if (validateOptions && !isEmptyObject(validateOptions)) {
-      fieldsWithValidationRef.current.add(name);
+      if (!validationFields || validationFields.includes(name))
```
Comment (barrymay) [inline src/useForm.ts]: minor code style issue (we can agree to disagree) all if statements should have brackets, event single line ones (this prevents accidental bugs if someone adds another line

Comment (bluebill1049) [reply src/useForm.ts]: agree :) fixing

Comment (barrymay) [reply src/useForm.ts]: cool, pretty sure there's an eslint rule for that if you IDE isn't reporting that

**Comment by barrymay** [comment]
general question - does this change still allow the case for onSubmit where you can click submit and validate regardless of errors for the first time?  I can confirm it with a beta cut, but if you've tested this, then all good.

File: src/useForm.ts (hunk 1 above)
Comment (bluebill1049) [reply src/useForm.ts]: I think my IDE doesn't report that at the moment (webstorm)

Comment (barrymay) [reply src/useForm.ts]: ok, I'll check into it

**Comment by bluebill1049** [comment]
> general question - does this change still allow the case for onSubmit where you can click submit and validate regardless of errors for the first time? I can confirm it with a beta cut, but if you've tested this, then all good.

This change only affect `isValid` form state, it should consistently return false at the initial render. I did test it with example and automation.

**Comment by barrymay** [comment]
cool - playing devil's advocate here... what if the starting state of the form was actually valid.  So my general question is ( and it might need more work to fully fix ) .  Should isValid be undefined on a form until the first validation turn is taken?

**Review by barrymay — APPROVED** [review]
TBH, I'm not sure what exactly this effects, but I don't see a reason to block it (it's more likely false than true to be sure)

**Comment by bluebill1049** [comment]
> cool - playing devil's advocate here... what if the starting state of the form was actually valid. So my general question is ( and it might need more work to fully fix ) . Should isValid be undefined on a form until the first validation turn is taken?

That's perfect! we can improve on this if users report such use case.

File: src/useForm.ts (hunk 1 above)
Comment (JeromeDeLeon) [reply src/useForm.ts]: [eslint curly](https://eslint.org/docs/rules/curly)

**Review by JeromeDeLeon — APPROVED** [review]
LGTM! I haven't been able to catch up lately. Sorry, guys. 😞

**Comment by JeromeDeLeon** [comment]
> cool - playing devil's advocate here... what if the starting state of the form was actually valid. So my general question is ( and it might need more work to fully fix ) . Should isValid be undefined on a form until the first validation turn is taken?

How do we affirm this? Do you have any example in mind? @barrymay 
I'm curious as well how do we handle such cases, but I think there will be rare cases.

**Comment by barrymay** [comment]
@JeromeDeLeon to be clear, I'm not blocking this change, but I'm curious how this came up and what problems it's causing, since isValid is not used as an exposed value of useForm.  

@bluebill1049 do you have a use case where this caused an issue?  I see the Cypress changes (great) but I don't understand why this suddenly is an issue.

File: src/useForm.ts (hunk 1 above)
Comment (barrymay) [reply src/useForm.ts]: thanks - please see #391 ;)

**Comment by bluebill1049** [comment]
@barrymay one of the issues, people reported for inconsistency for `isValid` which i agree

current behaviour

|  mode | isValid  |
|---|---|
|  submit | true  |
|  change | false  |
| blur  |  false  |

after this fix

|  mode | isValid  |
|---|---|
|  submit | false  |
|  change | false  |
| blur  |  false  |

also validationFields is not correctly reflected on `isValid` attribute as well, it shouldn't be included during valid check when it excludes such fields

**Comment by barrymay** [comment]
well... it depends what isValid is used for (if you can point to one of those issues raised or an email you got about it, I'd be happy to see the relationship)  ;) In my case I'm not using it so this change doesn't bother me much. 

I still feel that the undefined status of isValid would be better until the initial validation run is performed, but that will admittedly take some work to get right.  

Feel free to merge :)

**Comment by barrymay** [comment]
btw, is there a beta with this in it?

**Comment by bluebill1049** [comment]
there is a beta: `3.24.2-beta.2`

the first one: 
https://github.com/react-hook-form/react-hook-form/issues/298#issuecomment-543571586

the second issue: is reported in spectrum with validationFields  
https://spectrum.chat/react-hook-form/help/best-way-to-register-unregister-conditional-form-fields~a86951dd-3a86-4d00-8c99-b72a2e37c7b4

**Comment by barrymay** [comment]
Thanks - I’ll test and revert back with results

**Comment by barrymay** [comment]
all good - let's go with it.  thanks!

**Comment by bluebill1049** [comment]
thanks @barrymay for verifying.

**Pull request merged**